### PR TITLE
Frank/pathbased pubsub

### PR DESCRIPTION
**Pull request by frankosterfeld**
Still missing: SUB support in client class, more advanced query-based topic matching

### CODE REVIEW

File: src/majordomo/include/majordomo/SubscriptionMatcher.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+        if (pathSubscriberView.ends_with("*")) {
+            pathSubscriberView.remove_suffix(1);
+        }
+
+        // match path (leading characters) only - assumes trailing asterisk
+        return pathNotification->starts_with(pathSubscriberView);
+    }
+};
+
```
Comment (RalphSteinhagen) [inline src/majordomo/include/majordomo/SubscriptionMatcher.hpp:52]: `testPathOnly(...)` -> missing `testPathAndContext(...)` 

should be address now or latest once the [Majordomo]Worker with the necessary context `<C>` definitions is available
Two options:
  a) min-impl: matching of exact <key[=<value] pairs of notify and subscription topic (N.B. 'key' being case-insensitive, the 'values' not (necessarily), N.B 'values' being optional)
  b) implement 'TimingCtx'  and `Filter` interface/concept (tbd.)

File: src/majordomo/test/subscriptionmatcher_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    REQUIRE(matcher(URI("/property?testKey1=42&testKey2=24"), URI("/property?testKey2=24&testKey1=42")));
+    REQUIRE_FALSE(matcher(URI("/property?testKey1=41"), URI("/property*?testKey1=4711")));
+    REQUIRE(matcher(URI("/property/A?testKey1=41"), URI("/property*?testKey1=41")));
+    REQUIRE(matcher(URI("/property"), URI("/property?testQuery"))); // ignore unknown ctx filter on subscription side
+    REQUIRE(matcher(URI("/property"), URI("/property*?testQuery"))); // ignore unknown ctx filter on subscription side
+    REQUIRE(matcher(URI("/property?testKey1"), URI("/property*")));
+    REQUIRE(matcher(URI("/property?testKey1"), URI("/property?TestKey1"))); // N.B. key is case sensitive
+    REQUIRE_FALSE(matcher(URI("/property?testKey1"), URI("/property?testKey1=42")));
+    REQUIRE_FALSE(matcher(URI("/property"), URI("/property?testKey1=42")));
+}
```
Comment (RalphSteinhagen) [inline src/majordomo/test/subscriptionmatcher_tests.cpp:48]: to note: cross-check w.r.t. Java-based [unit-test](https://github.com/fair-acc/opencmw-java/blob/d1a0cc795ff4d3b3369bb374f27190bd6c629895/core/src/test/java/io/opencmw/filter/SubscriptionMatcherTest.java) cases

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp [hunk 3]
```diff
@@ -35,36 +35,42 @@ concept HandlesRequest = requires(T handler, RequestContext &context) { std::inv
...
 public:
-    explicit BasicMdpWorker(std::string_view serviceName, URI brokerAddress, RequestHandler &&handler, const Context &context = {}, Settings settings = {})
-        : _handler{ std::forward<RequestHandler>(handler) }, _settings{ std::move(settings) }, _brokerAddress{ std::move(brokerAddress) }, _serviceName{ std::move(serviceName) }, _context(context) {
-    }
-
-    explicit BasicMdpWorker(std::string_view serviceName, URI brokerAddress, RequestHandler &&handler, Settings settings)
-        : BasicMdpWorker(serviceName, brokerAddress, std::forward<RequestHandler>(handler), {}, settings) {
+    explicit BasicMdpWorker(std::string_view serviceName, EndpointURI brokerAddress, RequestHandler &&handler, const Context &context, Settings settings = {})
+        : _handler{ std::forward<RequestHandler>(handler) }, _settings{ std::move(settings) }, _brokerAddress{ std::move(brokerAddress) }, _serviceName{ std::move(serviceName) }, _context(context), _notifySocket(_notifyContext, ZMQ_PAIR), _notifyListenerSocket(_notifyContext, ZMQ_PAIR) {
+        static std::atomic<int> idCounter     = 0;
```
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:65]: for example, [see](https://github.com/fair-acc/opencmw-cpp/blob/0be90a0143e03e7b06ea79fd6423173b21ab4f66/src/majordomo/include/majordomo/RestBackend.hpp#L47-L75):
```cpp
    struct Worker {
        Socket         socket;
        zmq_pollitem_t pollItem{};


        Worker(const Context &context, const std::string &address)
            : socket(context, ZMQ_DEALER) {
            pollItem.socket = socket.zmq_ptr;
            pollItem.events = ZMQ_POLLIN;


            zmq_invoke(zmq_connect, socket, address).assertSuccess();
        }
    };


    std::unordered_map<std::thread::id, Worker> _workers;
    std::mutex                                  _workersLock;


    // Function that returns this thread's worker instance
    Worker &workerForThisThread() {
        auto threadId = std::this_thread::get_id();
        {
            std::lock_guard lock{ _workersLock };
            if (!_workers.contains(threadId)) {
                _workers.emplace(std::piecewise_construct,
                        std::forward_as_tuple(threadId),
                        std::forward_as_tuple(_context, URI<>::factory(_address).toString()));
            }
            return _workers.at(threadId);
        }
    }
```

**Review by RalphSteinhagen — CHANGES_REQUESTED** [review]
Looks good, two mains things to be addressed:
 a) decoupling of `notify(MdpMessage)` interface similar to the REST-Broker decoupling
 b) SubscriptionMatcher checks only for paths, option
    b.1) min-impl: matching of exact <key[=<value] pairs of notify and subscription topic
    b.2) implement the `TimingCtx` together with the [Majordomo]Worker impl.

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp [hunk 4]
```diff
@@ -35,36 +35,42 @@ concept HandlesRequest = requires(T handler, RequestContext &context) { std::inv
...
     Timestamp                     _heartbeatAt;
-
-    const Context &               _context;
-    std::optional<Socket>         _socket;
-    std::array<zmq_pollitem_t, 1> _pollerItems;
+    const Context                &_context;
+    std::optional<Socket>         _workerSocket;
+    std::optional<Socket>         _pubSocket;
+    std::array<zmq_pollitem_t, 3> _pollerItems;
+    const Context                 _notifyContext;
```
[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:57]: Wouldn't that fail if context is used tcp:/, while notify always uses inproc?

[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:57]: actually, the whole API doesn't allow for broker<->worker pubsub via tcp... but connecting the notify sockets fails when trying to reuse the broker context.

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp [hunk 5]
```diff
@@ -80,35 +86,54 @@ class BasicMdpWorker {
         _shutdownRequested = true;
     }
 
+    void notify(std::string topic, std::string body) {
```
[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:89]: Done, although most fields of the message are overwritten, only topic, body, error remain.

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp [hunk 6]
```diff
@@ -80,35 +86,54 @@ class BasicMdpWorker {
...
+        auto notify = createMessage(Command::Notify);
+        notify.setTopic(std::move(topic), MessageFrame::dynamic_bytes_tag{});
+        notify.setBody(std::move(body), MessageFrame::dynamic_bytes_tag{});
+        notify.send(_notifySocket).isValid();
+    }
+
+    void notifyError(std::string topic, std::string error) {
```
[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:96]: Done

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp [hunk 7]
```diff
@@ -80,35 +86,54 @@ class BasicMdpWorker {
...
     }
 
     void run() {
-        if (!connectToBroker())
+        if (!connectToBroker()) {
             return;
+        }
 
-        assert(_socket);
+        assert(_workerSocket);
```
[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:115]: these are checking internal invariants, not valid error condituions. if connectToBroker succeeds, the socket must be set.

File: src/majordomo/include/majordomo/SubscriptionMatcher.hpp (hunk 1 above)
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/SubscriptionMatcher.hpp:52]: I'd do that separately, as the design of the Filters might depend also on the serialisation. (I have a TimingCtx class now, I could do a first simple version quickly)

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (RalphSteinhagen) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp:109]: I agree that these are useful during development. Once these internal invariants are established (and never fail) they are basically dead-code and could and should be removed -> 'zero waste' principle

File: src/majordomo/include/majordomo/SubscriptionMatcher.hpp (hunk 1 above)
Comment (RalphSteinhagen) [reply src/majordomo/include/majordomo/SubscriptionMatcher.hpp:51]: There is already the refl-cpp solution for serialisation using the 
```cpp
ENABLE_REFLECTION_FOR(className, field1, field2, field3)
```
visitor pattern. If the `ENABLE_REFLECTION_FOR` is defined, then one can loop through all member fields and methods. using the
```cpp
refl::util::for_each(refl::reflect(value).members, [&](const auto member, [[maybe_unused]] const auto index) {
 // [..]
}
```
looping of the members. Have a look at the `opencmw::serialise(IoBuffer &buffer, const T &value, const uint8_t hierarchyDepth)` or [refl-cpp](https://github.com/veselink1/refl-cpp) directly for usage examples.

File: src/majordomo/test/subscriptionmatcher_tests.cpp (hunk 2 above)
Comment (RalphSteinhagen) [reply src/majordomo/test/subscriptionmatcher_tests.cpp:48]: The Java test-cases also includes the query parameters (or ignoring thereof if it's path-only checking)

Comment (frankosterfeld) [reply src/majordomo/test/subscriptionmatcher_tests.cpp:48]: That's the"if 0"'d test below. Then there's another one using TimingCtx and mimetypes, which I didn't include here yet.

File: src/majordomo/include/majordomo/BasicMdpWorker.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (frankosterfeld) [reply src/majordomo/include/majordomo/BasicMdpWorker.hpp]: I added another commit removing the internal checks

**Review by RalphSteinhagen — APPROVED** [review]

**Pull request merged**